GCEWindowsAgent/serial_listener.go

```diff
@@ -0,0 +1,462 @@
+	gsRegex     = regexp.MustCompile(fmt.Sprintf(`^gs://%s/%s$`, bucket, object))
+	// Check for the Google Storage URLs:
+	// http://<bucket>.storage.googleapis.com/<object>
+	// https://<bucket>.storage.googleapis.com/<object>
+	gsHTTPRegex1 = regexp.MustCompile(fmt.Sprintf(`^http[s]?://%s\.storage\.googleapis\.com/%s$`, bucket, object))
+	// http://storage.cloud.google.com/<bucket>/<object>
+	// https://storage.cloud.google.com/<bucket>/<object>
+	gsHTTPRegex2 = regexp.MustCompile(fmt.Sprintf(`^http[s]?://storage\.cloud\.google\.com/%s/%s$`, bucket, object))
+	// Check for the other possible Google Storage URLs:
+	// http://storage.googleapis.com/<bucket>/<object>
+	// https://storage.googleapis.com/<bucket>/<object>
+	//
+	// The following are deprecated but checked:
+	// http://commondatastorage.googleapis.com/<bucket>/<object>
+	// https://commondatastorage.googleapis.com/<bucket>/<object>
+	gsHTTPRegex3 = regexp.MustCompile(fmt.Sprintf(`^http[s]?://(?:commondata)?storage\.googleapis\.com/%s/%s$`, bucket, object))
+)
+
+func clearCurrentMsg() {
+	currentMsg = nil
+}
+
+func parseMessage() *serialRequest {
+	defer clearCurrentMsg()
+	msg := serialRequest{}
+	err := json.Unmarshal(currentMsg, &msg)
+	if err != nil {
+		logger.Fatalf(err.Error())
+	}
+	return &msg
+}
+
+func handlePrepareForSnapshotRequest(req serialRequest) {
+	if seenPreSnapshotoperationIds.Contains(req.OperationId) {
+		logger.Infof("ignoring pre-snapshot request with operation id %d", req.OperationId)
+		return
+	}
+
+	seenPreSnapshotoperationIds.Add(req.OperationId, nil)
+
+	logger.Infof("running pre-snapshot script")
+	config := getConfig()
+	code := runScript(config.PreSnapshotScriptUrl, config.Timeout)
+	res := serialResponse{identifier, "PRERESP", req.Version, code, req.OperationId}
+	out, err := json.Marshal(res)
+	if err != nil {
+		logger.Fatalf(err.Error())
+	}
+	writeSerial(serialPort, out)
+}
+
+func handleResumePostSnapshotRequest(req serialRequest) {
+	if seenPostSnapshotoperationIds.Contains(req.OperationId) {
+		logger.Infof("ignoring post-snapshot request with operation id %d", req.OperationId)
+		return
+	}
+	seenPostSnapshotoperationIds.Add(req.OperationId, nil)
+
+	logger.Infof("running post-snapshot script")
+	config := getConfig()
+	code := runScript(config.PostSnapshotScriptUrl, config.Timeout)
+	res := serialResponse{identifier, "POSTRESP", req.Version, code, req.OperationId}
+	out, err := json.Marshal(res)
+	if err != nil {
+		logger.Fatalf(err.Error())
+	}
+	writeSerial(serialPort, out)
+}
+
+func handleMessage(msg serialRequest) {
+	switch sig := msg.Signature; sig {
+	case "PREREQ":
+		handlePrepareForSnapshotRequest(msg)
+	case "POSTREQ":
+		handleResumePostSnapshotRequest(msg)
+	default:
+		logger.Fatalf("unknown message signature %s", msg.Signature)
+	}
+}
+
+func parseSerialData(serialData []byte) {
+	for _, c := range serialData {
+		if c == '\n' {
+			// Should be the end of the message
+			handleMessage(*parseMessage())
```

Comment:
i call parsemessage (on the global), then i dereference its return, and hand that to handlemessage? this feels janky